fathom/src/core/semantics.rs

```diff
@@ -330,6 +330,107 @@ impl<'arena, 'env> EvalContext<'arena, 'env> {
     }
 }
 
+/// Primitive evaluation step.
+type PrimStep =
+    for<'arena> fn(&ElimContext<'arena, '_>, &[Elim<'arena>]) -> Option<ArcValue<'arena>>;
+
+macro_rules! step {
+    ($context:pat, [$($input:pat),*] => $output:expr) => {
+        Some(|$context, spine| match spine {
+            [$(Elim::Fun($input)),*] => Some($output),
+            _ => return None,
+        })
+    };
+}
+
+macro_rules! const_step {
+    ([$($input:ident : $Input:ident),*] => $output:expr) => {
+        step!(_, [$($input),*] => match ($($input.as_ref(),)*) {
+            ($(Value::Const(Const::$Input($input)),)*) => Arc::new(Value::Const($output)),
+            _ => return None,
+        })
+    };
+}
+
+/// Returns an evaluation step for a primitive, if there is one defined.
+fn prim_step(prim: Prim) -> Option<PrimStep> {
+    use std::ops::{BitAnd, BitOr, BitXor, Not, Shl, Shr};
+
+    match prim {
+        Prim::FormatRepr => step!(context, [format] => context.apply_repr(format)),
+
+        Prim::U8Add => const_step!([x: U8, y: U8] => Const::U8(u8::checked_add(*x, *y)?)),
+        Prim::U8Sub => const_step!([x: U8, y: U8] => Const::U8(u8::checked_sub(*x, *y)?)),
+        Prim::U8Mul => const_step!([x: U8, y: U8] => Const::U8(u8::checked_mul(*x, *y)?)),
+        Prim::U8Div => const_step!([x: U8, y: U8] => Const::U8(u8::checked_div(*x, *y)?)),
+        Prim::U8Not => const_step!([x: U8] => Const::U8(u8::not(*x))),
+        Prim::U8Shl => const_step!([x: U8, y: U8] => Const::U8(u8::shl(*x, *y))), // TODO: Use `checked_shl`
+        Prim::U8Shr => const_step!([x: U8, y: U8] => Const::U8(u8::shr(*x, *y))), // TODO: Use `checked_shr`
```

Comment:
Need to decide on the types I expect here. Rust seems to expect `u32`s on the rhs of [`checked_shl`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_shl) and [`checked_shr`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_shr)?

Reply:
`u32` allows for a lot of shifts!

Reply:
I know! :O

Reply:
is `U8` ok? Surely? 🤔

Reply:
"`U8` ought to be enough for anybody"

Reply:
Might just do some conversions